### PR TITLE
Implement the CONVERT_TZ() SQL function

### PR DESCRIPTION
**Pull request by VinaiRachakonda**

### CODE REVIEW

File: sql/expression/function/convert_tz.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	datetime, err := c.dt.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	datetimeStr, ok := datetime.(string)
```
Comment (zachmu) [inline sql/expression/function/convert_tz.go:82]: This isn't right. This value can be anything that can be converted to a datetime, not just a string

Comment (zachmu) [reply sql/expression/function/convert_tz.go:82]: Need a test for this

**Review by zachmu — COMMENTED** [review]
What's here is good, but this needs to work on datetime columns as wel

**Comment by VinaiRachakonda** [comment]
Great catch

File: sql/datetimetype.go [hunk 2]
```diff
@@ -205,7 +205,7 @@ func (t datetimeType) ConvertWithoutRangeCheck(v interface{}) (time.Time, error)
 			return zeroTime, ErrConvertingToTime.New(v)
 		}
 	case time.Time:
-		res = value.UTC()
+		res = value
```
Comment (VinaiRachakonda) [inline sql/datetimetype.go:208]: @zachmu Pretty confident that we need this change. For datetime objects (time.Time) it's essential we return the object with the correct timezone loaded in. Converting to UTC consistently does not seem correct

**Review by zachmu — COMMENTED** [review]
Other changes look fine, want Daylon's opinion on the UTC semantics change

File: sql/datetimetype.go (hunk 2 above)
Comment (Hydrocharged) [reply sql/datetimetype.go:208]: I actually don't think this will change anything for any tests that hit Dolt. All throughout Dolt we call `time.UTC()` to _make sure_ that we're standardizing them. There's probably other places in gms where we do this `time.UTC()` call so I'm not sure if this is even changing anything. From a conceptual level though, this particular change affects both `TIMESTAMP` and `DATETIME`. `TIMESTAMP` is definitely supposed to convert to UTC, and `DATETIME` does not.

For the purposes of `CONVERT_TZ`, this change shouldn't be necessary. When you call `time.UTC()`, Go is simply setting the internal location to UTC. It's easy to simply set it back in your `CONVERT_TZ` function (which you're going to have to do anyway). It doesn't change the internal unix timestamp at all. In fact, all locations/timezones are just for display purposes as far as Go is concerned, so converting it to UTC, finding the timezone delta, and applying that to the time gives the exact same result as far as what is exposed from MySQL.

We shouldn't be comparing times using `==`, but if we are, having everything as UTC at least ensures that all of those will evaluate correctly. We could treat UTC as "no timezone", and any "timezone modifications" would just offset the actual timestamp by some amount. There are probably more changes that would need to be made to accommodate this, but it's far easier if we can ensure all times are set to UTC to begin with.

TL;DR: this change isn't necessary and we can leave it as UTC() for the sake of consistency, as multiple places reference this. If this change is required to make your code work, then I'm not sure that code is correct.

File: sql/expression/function/convert_tz.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	// If either the date, or the timezones/offsets are not correct types we return NULL.
+	datetime, err := sql.Datetime.ConvertWithoutRangeCheck(dt)
+	if err != nil {
+		return nil, nil
+	}
+
+	// Get the time in UTC.
+	datetime = time.Date(datetime.Year(), datetime.Month(), datetime.Day(), datetime.Hour(), datetime.Minute(), datetime.Second(), datetime.Nanosecond(), time.UTC)
```
Comment (Hydrocharged) [inline sql/expression/function/convert_tz.go:89]: If we remove this, does anything change? `ConvertWithoutRangeCheck` should already be returning a UTC time, so this seems like it shouldn't change anything.

File: sql/expression/function/convert_tz.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	tLoc, err := time.LoadLocation(toLocation)
+	if err != nil {
+		return time.Time{}
+	}
+
+	getCopy := func(t time.Time, loc *time.Location) time.Time {
+		return time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), loc).UTC()
+	}
```
Comment (Hydrocharged) [inline sql/expression/function/convert_tz.go:129]: I'd make this a function in the module namespace rather than one created inside of this function. I'm assuming this was done to avoid cluttering the namespace, so you could always make it a function of `*ConvertTz`, e.g.:
```go
func (c *ConvertTz) getCopy(t time.Time, loc *time.Location) time.Time {
	return time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), loc).UTC()
}
```

File: sql/expression/function/convert_tz.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		return nil, nil
+	}
+
+	toStr, ok := to.(string)
+	if !ok {
+		return nil, nil
+	}
+
+	converted := convertTimeZone(datetime, fromStr, toStr)
+	if !converted.IsZero() {
```
Comment (Hydrocharged) [inline sql/expression/function/convert_tz.go:102]: If `converted` is the zero time then we attempt to parse offsets. What if the conversation is supposed to result in the zero time? Is that an error or is that valid? Perhaps a `success` bool being returned if it is valid.

**Review by Hydrocharged — APPROVED** [review]
Looks good! A couple of things and you're free to merge!

File: sql/expression/function/convert_tz.go (hunk 4 above)
Comment (VinaiRachakonda) [reply sql/expression/function/convert_tz.go:129]: Not sure if it needs to be a method on ConvertTz. I'll take this out of the other function though
